### ISSUE TITLE
problem in onUse() [TFS 1.1.x]

### ISSUE THREAD

**Issue by whitevo** [otland/forgottenserver#1572]
if you use item what is made so it cant be moved:
function Player:onMoveItem(item, count, fromPosition, toPosition)
if item:getId() == 2674 then return false end
end

And you try to use it on itemEx, what is not reachable.
Then first you pick up the item and then messages comes.. cant be reached.
But now you can't move that item out of your bag. because well.. it was supposed to be unmoveable item..

How to fix that?

**Comment by Tarliton** [comment]
i can confirm this:
as far as i could check this happens because of this:
the byte that the server recieves when a player try to move an item is 0x78 in that case the server uses the function parseThrow(msg) to handle the player item move. Somewhere in that function call there is the check for event:

if (!g_events->eventPlayerOnMoveItem(player, item, count, fromPos, toPos)) {
               return;
     }

this check prevent the item from being moved if you activate( like whitevo did).

But, when you try to use an item (use with...) the server recieves the byte 0x83 and uses the function parseUseItemEx(msg) to handle it. In that function the item is moved iternally (not by Game::playerMoveItem) and i could not find the check for eventPlayerOnMoveItem to deny if you decided to not move the said item.
So i was able to fix this by adding the g_events->eventPlayerOnMoveItem check inside the Game::playerUseItemEx but i dont know if this is a good way to fix this issue.

**Comment by whitevo** [comment]
How you create the check?
Like this?

```
void Game::playerUseItemEx(uint32_t playerId, const Position& fromPos, uint8_t fromStackPos, uint16_t fromSpriteId,
                           const Position& toPos, uint8_t toStackPos, uint16_t toSpriteId)
{
    if (! g_events->eventPlayerOnMoveItem()) {
		return;
	}
```
I have now clue how C++ works..

**Comment by Tarliton** [comment]
in function Game::playerUseItemEx file game.cpp:
bellow line that contains:
```c++
if (fromPos.x != 0xFFFF && toPos.x != 0xFFFF && Position::areInRange<1, 1, 0>(fromPos, player->getPosition()) &&
			        !Position::areInRange<1, 1, 0>(fromPos, toPos)) {
				Item* moveItem = nullptr;
```

paste:
```c++
if (!g_events->eventPlayerOnMoveItem(player, item, item->getItemCount(), fromPos, toPos)) {
					player->sendCancelMessage(RETURNVALUE_CANNOTPICKUP);
					return;
				}
```

**Comment by whitevo** [comment]
aah haven't come to github for so long xD

I think I used your method Tarliton and it worked perfectly.

You can close this thread.

**Comment by Tarliton** [comment]
hi, is that the proper way of solving this? 
if so I could open a pull request.
thanks

**Comment by marksamman** [comment]
It is, for the case described in this issue. There are more cases where this should be handled, maybe we should just move the event call into internalMoveItem, but it might still be possible to have the item moved by trading it. This issue isn't a high priority or a release blocker for 1.2, so I haven't looked at it and probably won't do so before the 1.2 release, but if you'd like to submit a pull request that addresses all the cases I'll review your proposed solution.

**Comment by whitevo** [comment]
what is the function name for eventPlayerOnMoveItem for TFS 1.3?
Since this bug is still here, but don't know how to get return value from the lua function 
Player:onMoveItem(item, count, fromPosition, toPosition, fromCylinder, toCylinder)

**Comment by whitevo** [comment]
Ok seems the function exists, but I guess I need to add cylinders?
Well this is as much I gathered. Could I simply give nil values somehow if its not possible to create cylinders.

`toCylinder = internalGetCylinder(player, toPos);
	if (toCylinder == nullptr) {
		player->sendCancelMessage(RETURNVALUE_NOTPOSSIBLE);
		return;
	}
	
	const Position& mapFromPos = fromCylinder->getTile()->getPosition();
	Cylinder* fromCylinder = internalGetCylinder(player, fromPos);
	
	if (fromCylinder == nullptr) {
		player->sendCancelMessage(RETURNVALUE_NOTPOSSIBLE);
		return;
	}
	
	const Tile* toCylinderTile = toCylinder->getTile();
	const Position& mapToPos = toCylinderTile->getPosition();
	
	if (!g_events->eventPlayerOnMoveItem(player, item, item->getItemCount(), fromPos, toPos, fromCylinder, toCylinder)) {
		player->sendCancelMessage(RETURNVALUE_CANNOTPICKUP);
		return;
	}`

**Comment by Zbizu** [comment]
Getting "sorry not possible" on master branch when trying with shovel (2554)

Can be closed I guess (unless we still support 1.1 branch).
